gazebosim/gz-sim
# New system plugin called `DriveToPoseController`

# 🎉 New feature

Closes #2379 

## Summary

This PR adds a new system plugin to `gz-sim` called `DriveToPoseController`.  This is not a standalone plugin and requires other systems plugins (`DiffDrive` and `OdometryPublisher`) to function.

This controller receives a target pose in the Gazebo coordinate system on `/cmd_pose` topic and calculates and publishes a command velocity on `/cmd_vel` topic to move the robot towards the target. This plugin contains separate linear and angular proportional controllers with adjustable gains. The user can also specify an allowable deviation from the goal pose in the form of linear and angular deviations. Upon reaching the goal pose, the controller publishes an acknowledgement on `/reached_pose` topic to indicate the end of the navigation operation.

The plugin needs to be attached to a model entity as in the SDF as follows,

```xml
    <model>
         
 
        

         
        <plugin
          filename="gz-sim-drive-to-pose-controller-system"
          name="gz::sim::systems::DriveToPoseController">
          <linear_p_gain>1.0</linear_p_gain>
          <angular_p_gain>2.0</angular_p_gain>
          <linear_deviation>0.1</linear_deviation>
          <angular_deviation>0.05</angular_deviation>
        </plugin>
    </model>
```

It has the following parameter tags,
- `<linear_p_gain>` - (Optional) Proportional gain for the linear velocity controller. Default: 1.0
- `<angular_p_gain>` - (Optional) Proportional gain for the angular velocity controller. Default: 2.0
- `<linear_deviation>` - (Optional) Allowable linear deviation (in meters) from the desired coordinate. Default: 0.1
- `<angular_deviation>` - (Optional) Allowable angular deviation (in radians) from the desired orientation. Default: 0.05

The following example warehouse world can be found at `src/gz-sim/examples/worlds/drive_to_pose_controller.sdf` with an AMR spawned at its start position.

![image](https://github.com/user-attachments/assets/0f52c802-ad16-4bfe-83d7-43a800e22db3)

![image](https://github.com/user-attachments/assets/3af0a3ef-9af3-4a39-896c-13795a329d53)

## Test it

The integration test for this plugin can be run using the following command,
```bash
  colcon test --event-handlers console_direct+ --merge-install \
    --packages-select gz-sim9 --ctest-args -R drive_to_pose_controller
``` 

Use the following steps to test the functionality in the warehouse world by sending the robot to four different poses,

1. Launch the following world, 

```bash
  gz sim drive_to_pose_controller.sdf
```
> NOTE: The world opens in a paused state so make sure to unpause before sending the pose commands

2. Send the robot to the first pose in the warehouse, `[x, y, orientation] = [1.78, -9.4, -3.14]`

```bash
  gz topic -t "/model/DeliveryBot/cmd_pose" -m gz.msgs.Pose_V \
    -p "pose:[{position: {x: 1.78, y: -9.4, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: -0.99, w: 0.0}}]"
```

3. Send the robot to the second pose, `[x, y, orientation] = [-3.75, -9.4, 1.57]`

```bash
  gz topic -t "/model/DeliveryBot/cmd_pose" -m gz.msgs.Pose_V \
    -p "pose:[{position: {x: -3.75, y: -9.4, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.707, w: 0.707}}]"
```

4. Send the robot to its third pose, `[x, y, orientation] = [-3.75, 9.18, 3.14]`

```bash
  gz topic -t "/model/DeliveryBot/cmd_pose" -m gz.msgs.Pose_V \
    -p "pose:[{position: {x: -3.75, y: 9.18, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.99, w: 0.0}}]"
```

5. Send the robot to its final pose, `[x, y, orientation] = [-5.9, 9.18, 3.14]`

```bash
  gz topic -t "/model/DeliveryBot/cmd_pose" -m gz.msgs.Pose_V \
    -p "pose:[{position: {x: -5.9, y: 9.18, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.99, w: 0.0}}]"
```

The entire operation is shown in the following video,

https://github.com/user-attachments/assets/84b1024a-2a78-4bfc-9bc3-ac2fa0f50d62

## Checklist

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.
